ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java

```diff
@@ -768,6 +757,25 @@ private final void shutdown(boolean waitForTransactions)
         _reaperWorkerThread = null;
     }
 
+    /**
+     * Wait for all the transactions to terminate normally.
+     *
+     * Note: this method assumes that the transaction system has been
+     * shutdown already so no new transactions can be created, or we
+     * could be here for a long time!
+     */
+    public void waitForAllTxnsToTerminate() {
```

Comment:
Is there some way to test that the transaction system has been shutdown already in this method and throw an exception if not?

What should be the behaviour if TxControl is enabled part the way through this shutdown I suppose?

Reply:
By us not knowing the terminate has been called, we could be in here a long time even if no new transactions are created. (e.g timeout in 1y). Terminate can set to wait for the transactions to false (which looks like it does wait, but it sets their timeout to zero)?

Reply:
> Is there some way to test that the transaction system has been shutdown already in this method and throw an exception if not?
> 
> What should be the behaviour if TxControl is enabled part the way through this shutdown I suppose?

The note in the javadoc is in line with all other (similar) methods in the Transaction Reaper. To keep the overall design of this class coherent, I would prefer to avoid introducing a check based on `TxControl.isEnalbed()` in this method.

Reply:
I guess we call terminate in WildFly somewhere (at least indirectly). Will get a reference and share here (basically supposing we might want the same)

Reply:
We don't seem to shut down the reaper. I guess we just rely on it being a daemon thread: https://github.com/jbosstm/narayana/blob/b8fc783bca6359ddf6641cd8f478c30595cf3d6e/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java#L905

Reply:
> > Is there some way to test that the transaction system has been shutdown already in this method and throw an exception if not?
> > What should be the behaviour if TxControl is enabled part the way through this shutdown I suppose?
> 
> The note in the javadoc is in line with all other (similar) methods in the Transaction Reaper. To keep the overall design of this class coherent, I would prefer to avoid introducing a check based on `TxControl.isEnalbed()` in this method.

Understood - thank you

Reply:
Thank you, Tom

Reply:
I think we can solve this thread. Thank you